mpremote: Add automatic PTY device detection for QEMU

andrewleech opened the pull request:
### Summary

This change addresses hangs when mpremote and pyboard.py communicate with PTY devices, particularly QEMU serial ports. PTY devices' `inWaiting()` method returns timing-dependent results and can report zero bytes available even when data is buffered. This causes mpremote/pyboard.py to wait indefinitely for data that will never be indicated as available.

The fix implements PTY device detection (Linux Unix98 PTY path `/dev/pts/N` with major device number 136) and uses blocking reads with pyserial's `interCharTimeout` for these devices. For regular serial devices, the existing `inWaiting()` polling behavior is preserved.

This change originated from OpenMV's downstream patch (https://github.com/openmv/openmv/blob/master/tools/mpremote-qemu-serial.patch) that resolved similar issues with QEMU testing.

### Testing

Tested on QEMU MPS2-AN385 port with stress test (3 iterations of 50-second sleep loops, 150 seconds total):
- With fix: completed successfully
- Without fix: hung after 90 seconds

Tested with real-world filesystem workload (12 PPM/PGM image files, 512KB heap):
- With fix: all files processed in 1m59s
- Without fix: hung on 4th file after 10 minutes

OpenMV project reports all 81 tests now pass on QEMU.

Also fixed `execpty:` device type handling in pyboard.py (used by QEMU CI tests) which was incorrectly marked as non-PTY, causing 6-hour CI hangs.

### Trade-offs and Alternatives

The PTY detection uses Linux-specific major device number check (136) and Unix98 PTY path pattern. Detection failures fall back to standard serial behavior, so non-Linux platforms or unusual configurations remain functional but may still experience PTY timing issues.

Blocking reads add 1-second `interCharTimeout` per byte on PTY devices. This is acceptable since PTY communication is typically local with low latency. The timeout ensures responsiveness if the remote end becomes unresponsive.

andrewleech commented:
@iabdalkader this was one of my distractions that got in the way of updating the openmv unix port PR - I'd appreciate a review in the hope it'll fix the issue you currently patch for in https://github.com/openmv/openmv/blob/master/tools/mpremote-qemu-serial.patch?

iabdalkader commented:
> I'd appreciate a review

I don't much about `mpremote` internals to properly review it, but I did test it and can confirm it fixes the issue. Normally, and without my patch, tests that read files could take up to 20 or 30 seconds each (that is if they don't fail completely). With this patch, those tests run in 2-3 seconds similar to my patch that bypasses inwaiting. The total time for about ~80 tests is 60 seconds, Vs. 7 *minutes*

FWIW I had a test case bundle here that reproduces the issue: https://github.com/micropython/micropython/pull/18234#issuecomment-3396284256

Note, it seems to make the CI tests here hang maybe.

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -140,22 +165,27 @@ def read_until(
+            # PTY: always read (blocking with timeout), Serial: check inWaiting() first
+            if self.is_pty or self.serial.inWaiting() > 0:
                 new_data = self.serial.read(1)
-                if data_consumer:
```
To make this diff smaller and easier to understand/review/test, how about this:
```py
new_data = None
if self.is_pty or self.serial.inWaiting() > 0:
    new_data = self.serial.read(1)
if new_data:
    ... (existing code, including the else)
```

pi-anl committed:
tools/{mpremote,pyboard.py}: Add PTY device detection for QEMU.

On PTY devices (e.g. QEMU serial), ioctl(FIONREAD) can race with
write() causing inWaiting() to return 0 when data is available.
This adds automatic PTY detection on Linux (/dev/pts/* with major
number 136) and skips the inWaiting() check for PTY devices,
falling through to the blocking serial.read() instead.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dpgeorge commented on tools/pyboard.py:
```diff
@@ -333,6 +334,27 @@ def __init__(
+            self.is_pty = True
+        elif device.startswith("exec:") or (
+            device and device[0].isdigit() and device[-1].isdigit() and device.count(".") == 3
+        ):
```
I don't think this elif condition is needed at all.  The check in `_is_pty_device()` will never match anything starting with `exec:` or looking like an IP address, so this elif can be removed.

dpgeorge commented on tools/pyboard.py:
```diff
@@ -333,6 +334,27 @@ def __init__(
+    def _is_pty_device(device):
+        """Detect if device is a PTY (pseudo-terminal), e.g. used by QEMU."""
+        try:
+            if device.startswith("/dev/pts/"):
```
Minor point, but for code style IMO the try should be inside the if (because you don't expect the if to fail with OSError/AttributeError).  Something like:
```py
if device.startswith("/dev/pts/"):
    try:
        st = os.stat(device)
        ...
    except (OSError, AttributeError):
        pass
return False
```

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -105,6 +105,20 @@ def __init__(self, device, baudrate=115200, wait=0, exclusive=True, timeout=None
+    def _is_pty_device(device):
+        """Detect if device is a PTY (pseudo-terminal), e.g. used by QEMU."""
+        try:
+            if device.startswith("/dev/pts/"):
```
Similar comment about code style as in `pyboard.py`.

dpgeorge reviewed (approved):
Thanks for updating, this looks good now.
